File: backend/dataset/maxPathSum/maxPathSum_20.py
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        _, m = self._maxPathSum(root)
        return m

    def _maxPathSum(self, tree):
        if tree == None:
            return (float('-inf'), float('-inf'))

        lsb, ls = self._maxPathSum(tree.left)
        rsb, rs = self._maxPathSum(tree.right)
        val = tree.val

        cb = max(lsb, rsb)
        b = max(cb + val, val)
        r = max(lsb + val + rsb, b)
        m = max(ls, rs, r)

        return (b , m)

File: backend/CDataset/maxPathSum/maxPathSum_20.cpp
class Solution {
public:
    class TreeNode{
        public:
        TreeNode* left;
        TreeNode* right;
        int val;
    };
    int maxPath( TreeNode* root ){
        if( !root )
            return 0;
        return max(root->val ,root->val + max(maxPath(root->left),maxPath(root->right) ));
    }
    int maxPathSum(TreeNode* root) {
            if( !root )
                return INT_MIN;
            return max(root->val + maxPath(root->right),max(root->val + maxPath(root->left),max(root->val,max(root->val + maxPath(root->left) + maxPath(root->right),max(maxPathSum(root->left),maxPathSum(root->right))))));
    }
};
